Fixes #3073 - send the correct command on reboot and reset

thoro opened the pull request:
<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

Currently pressing "Reboot" on the UI soft-shutdowns the server. See the referenced issue (https://projects.theforeman.org/issues/3073) for details. PR changes so that reboot reboots the server, and reset actually resets the server.

thoro committed:
Fixes #3073 - send the correct command on reboot and reset

ezr-ondrej commented:
ok to test

ezr-ondrej reviewed (dismissed):
Quite ancient bug, such a shame we did not notice earlier.
Thanks a bunch for the contribution @Thoro! :+1:
LGTM, but will wait for jenkins just in case :)

thoro commented:
Seems there's still something not correct, I have now updated it again, because the UI was still showing Cycle and Soft options (which are actually poweroff and reboot)

I also changed now poweroff, to soft poweroff, since there's also already the "Off" option to poweroff

thoro commented:
@ezr-ondrej I don't fully understand how the tests work for this specific case, and I have no experience with ruby regarding the unit tests - is there somewhere a local dev setup manual? (for mac os)

thoro commented:
I got now the testing locally working, but I had to modify some more, because it was doing some magic there about the supported actions. Not sure if this is how it should be in ruby - but that's how I got it working 😊

Would appreciate feedback here.

ezr-ondrej commented:
Thoro: https://theforeman.org/contribute.html#Runningtests here is quite a lot of information.
But I do not get what you're trying to achieve with those additional changes :thinking:

thoro commented:
@ezr-ondrej  The issue is that REAL_ACTIONS is used to fill the UI, so you got there Poweroff, Soft, Cycle etc. in the Host menu

So, what happened before:

- base.rb used SUPPORTED_ACTIONS to generate functions dynamically, which included soft,cycle,on,off - these functions forwarded the action directly to the proxy, since they were not defined in the action_map in bmc.rb
- bmc.rb defined action_map, so that `start` maps to `on`, `stop` maps to `off`, etc, since those functions are also defined in SUPPORTED_ACTIONS there is this akward if in the base.rb (i.e. either use send or default_action)
- this led to the tests failing, when I removed on,off,cycle,soft from the SUPPORTED actions (so that they are not displayed in the UI anymore)
- therefore I needed to change how that mapping happens, now instead of pointing to a different method, the default action hash value defines what is forwarded to the proxy (equals simple format), and the action hash value defines if another method is called.
- this allowed me to remove all mappings that have the same meaning

Can you follow why I did the changes?

ezr-ondrej commented:
> Can you follow why I did the changes?

Yeah, it make sense :) I'm off for the weekend, so I won't be able to review until monday and I'll have to read what we are really doing and how that works today to make sure we are not breaking something :)

I've run the rubocop now for you, but the foreman CI failure is because of an issue in rubocop and unfortunatelly I'd always need to run it manually for you, but you can see the output when you select one of the matrix versions and see the console output, eg: https://ci.theforeman.org/job/test_develop_pr_core/10175/database=postgresql,ruby=2.5,slave=fast/console

ezr-ondrej commented on app/services/power_manager/base.rb:
```diff
@@ -58,9 +56,11 @@ def translate_status(result)
       # if action_map is in simple format (:key => 'method'), transform it to entry
-      action_entry = {:action => action_entry.to_sym} unless action_entry.is_a?(Hash)
+      action_entry = {:default => action_entry.to_sym} unless action_entry.is_a?(Hash)
+      print "#{method} #{action_entry} \n"
```
we should not do this ^_^ ... probably a debug leftover, but if you really want to write to the logs, use [logger.debug](https://guides.rubyonrails.org/debugging_rails_applications.html#log-levels)

thoro replied:
yep debug left over 😊 - needed to understand during running the tests whats going on ...

ezr-ondrej commented on test/unit/power_manager_test.rb:
```diff
@@ -73,6 +73,7 @@ class PowerManagerTest < ActiveSupport::TestCase
   def actions_list(host)
     action_map = host.power.send(:action_map)
     actions = PowerManager::SUPPORTED_ACTIONS - action_map.keys.map { |k| k.to_s }
+  
```
Robocop doesn't like this line. Don't you have some spaces here? It should be empty line without any trailing white space.

thoro replied:
yep, there were spaces ... was doing the changes in vi - went over it again in Sublime now.

also tried Rubocop locally - and it found no issues anymore

ezr-ondrej replied:
Yep, Jenkins approves and the tests are green now :tada:

thoro commented:
@ezr-ondrej I have just realised that Reboot and Cycle are two different things.

I would expect Reboot to do a Graceful Shutdown and then immediately start the server again.
but cycle is an instant shutdown (like poweroff) and then start again.

Do you have any suggestion on how to solve this? Do we need to solve it?

Build also calls Reboot, not sure if a graceful shutdown is required there.

Edit:
possible solution would be:

Rename the Build option "Restart" to "Reset"
Rename "Reboot" to "Graceful Shutdown"
Introduce "Power Cycle" or "Cold Boot" to represent "Cycle"

Would mean that the reboot option goes away completely, but all the functions map correctly

ezr-ondrej commented:
I'm not the best person to judge this, I'll get back to you with an answer :)

lzap commented:
Our mapping is really weird. This is how IPMI should behave:

* on - turn on power
* off - cut power
* cycle - cut power, turn on power
* soft - soft power off via ACPI (regular shutdown as we like to do it)
* reset - hard reset of the BMC management board (can be warm or cold)

Mapping for both IPMI and Libvirt (examples for virsh commands) I would like to see and is safe and logical for 2021 (how I would like to see both Ruby symbols and UI strings):

* `:on` (Power on) - IPMI on - virsh start - Redfish On
* `:shutdown` (Safe shutdown) - IPMI soft - virsh shutdown - Redfish GracefulShutdown
* `:stop` (Force power off) - IPMI off - virsh destroy - Redfish ForceOff
* `:reset` (Force power reset) - IPMI cycle - virsh reset - Redfish ForceRestart
* `:mgmt_warm_reset` (Management warm reset) - IPMI reset - virsh N/A - Redfish Manager.Reset
* `:mgmt_cold_reset` (Management cold reset) - IPMI reset - virsh N/A - Redfish Manager.Reset

Note I propose to use "Force" word for all possibly dangerous actions. It would perhaps also make sense to indicate in the UI directly what IPMI commands these actually map to maybe via some tooltip or something. The question is if we should expose cold/warm IPMI management resets, Redfish does not support it, virtualization is irrelevant. Maybe we can just pick one or do a setting for that.

Man pages:

* https://linux.die.net/man/1/ipmitool
* https://linux.die.net/man/1/virsh

Edit: I think it is worth discussing this more widely before committing to this change: https://community.theforeman.org/t/rfc-change-how-bmc-operations-are-mapped/27241

lzap commented on app/services/power_manager.rb:
```diff
@@ -1,4 +1,4 @@
-  SUPPORTED_ACTIONS = REAL_ACTIONS + [N_('state'), N_('on'), N_('off'), N_('status'), N_('ready?')]
+  REAL_ACTIONS = [N_('start'), N_('stop'), N_('poweroff'), N_('reboot'), N_('reset')]
+  SUPPORTED_ACTIONS = REAL_ACTIONS + [N_('state'), N_('status'), N_('ready?')]
 end
```
What's the purpose of this hunk? Why are we dropping soft/cycle here?

lzap commented on app/services/power_manager/bmc.rb:
```diff
@@ -18,11 +18,9 @@ def action_map
+                         :poweroff => 'soft',
+                         :reboot   => 'cycle',
+                         :state    => { :action => 'status' },
                        })
```
I like poweroff=soft, makes sense. But reboot=cycle is very dangerous if we want to keep using the term we need to implement some kind of UI mapping for WebUI/CLI. But since this PR looks like a bigger refactoring, why not to completely rename those symbols to safe and meaningful variants?

lzap commented on app/services/power_manager/virt.rb:
```diff
@@ -35,12 +35,9 @@ def default_action(action)
+                         :poweroff => 'soft',
+                         :status   => {:action => :virt_state, :output => :state_output, :default => nil},
+                         :state    => {:action => :virt_state, :output => :state_output, :default => nil},
                        })
```
Don't understand why on/off/cycle is being dropped.

thoro replied:
it's just removed from the mapping table, because after the refactor the names match from the symbols in REAL_ACTION and the virt backend.

So:

start == start
stop == stop

that's why on / off is dropped, and cycle was my mistake to assume it's actually reboot!

=> but as you have provided already a pretty good expected behaviour I will go over it, rework what I did to match your proposal, and then we can take a look at it again.

thoro committed:
Fixes #3073 - implement proposed changes

thoro commented:
@lzap I added the proposed naming into a separate commit. I have also removed the duplicate state/status, now it's just status.

Is there anywhere the redfish implementation? I only saw virt and bmc!

lzap commented:
Apologies for the delay.

> Is there anywhere the redfish implementation? I only saw virt and bmc!

Redfish was initially implemented only recently, but it should work I tested it the other day. What exactly do you not see?

Let me re-review now.

lzap commented on app/services/power_manager/bmc.rb:
```diff
@@ -16,13 +16,13 @@ def ready?
       super.deep_merge({
-                         :start    => 'on',
+                         :on       => 'on',
                          :stop     => 'off',
```
Now that I look on it, it is inconsistent: on=on, stop=off. :) I have no preference but I would suggest to go either for on/off or start/stop.

lzap reviewed (changes requested):
Works fine, can you just fix the inconsistency I initially proposed? Also I would like to see those "Force" words for dangerous operations in the UI too, but this is not a must.

Please fix those mgmt strings, there must be some mapping that maps those short names into human-readable stings with spaces.

![Snímek obrazovky 2022-03-01 v 16 05 32](https://user-images.githubusercontent.com/49752/156193891-a8ed394b-62c0-4208-a080-c53b410fda3f.png)

ezr-ondrej commented:
@Thoro it was not missing much to get in, do you think you can finish it? Or let us know what you've struggled with so somone knows where to take over?

ezr-ondrej commented:
Reopening just in case there would be someone to finish it up :)

MariaAga commented:
Closing as for now no one can work on it.

spesnova717 commented:
@ezr-ondrej @lzap 

Hi,
I'm interested in contributing to Foreman. I have read contributor handbook.
I recently encountered this exact issue (#9082) myself. 
Since the issue still persists, I’d like to work on fixing it.
 
I’d like to ask:  
Would it be appropriate to continue the fix within this pull request (#9082).

Also, I noticed that there is a corresponding Redmine issue for this (Bug [#3073](https://projects.theforeman.org/issues/3073)).  
May I assign myself as the owner of that issue?
 
If there’s anything I should check or confirm in advance regarding how to proceed, I’d appreciate your guidance.

lzap commented:
Hirota-san, thank you for contacting us. We appreciate any help on this matter, so please feel free to take it from here and assign the issue and open a new PR. Link this PR from the new one please so possible reviewers will know the context as I am no longer maintaining this code but there will surely be someone from the core team for a review. The code looks good to me, I only requested few cosmetic changes but everything is already there. ありがとうございます

spesnova717 commented:
@lzap san
Thank you for the clarification.
I’ll organize the content on my side and submit a new pull request once it’s ready.
I’ll be sure to link the original PR for context.
Thank you again for your guidance.

spesnova717 commented:
Hi @lzap,
I tried to assign the [Redmine](https://projects.theforeman.org/issues/3073) issue to myself, but it seems I don’t have the necessary permission to do so.
Could you please assign it to me instead?
Thank you!

ezr-ondrej commented:
Hi @spesnova717
You might need to sign in to Redmine to get your account created. Only then we could assign you the ticket.
It's not as important to assign the ticket, I've left a comment in the ticket stating you're now working on this.
Let us know if you'd decided to get your account created.

lzap commented:
I can confirm I do not see your name in the list, I will ask folks from the core team why is that:

<img width="439" alt="image" src="https://github.com/user-attachments/assets/7bf2e0b5-6b49-4b97-b083-6085d2da8ac6" />

lzap commented:
And ... it is all set. Thank you, @spesnova717

spesnova717 commented:
@lzap @ezr-ondrej 
Thank you for assigning the ticket to me.
Also, thank you for the explanation and for leaving a note on the Redmine ticket.
I’ll proceed with preparing the updated pull request.
Appreciate your support!

spesnova717 commented:
@lzap san, @ezr-ondrej san

Hi,

Thank you again for your previous discussions on this topic.

Based on the prior work in this pull request and the current state of power operations in Foreman, I’ve reviewed the IPMI and Redfish specifications and drafted a proposal to align the supported power operations across these interfaces.

I’d greatly appreciate it if you could review the following proposal and let me know your thoughts.  
If this direction seems reasonable, I’d like to proceed with implementing it in a new pull request.

## Proposal for Restructuring Power Operations in Foreman

This document proposes a revision of the power control operations in Foreman.

---

### Background and Objective

After reviewing the official specifications of IPMI and Redfish, it became clear that only a limited set of **five basic power control functions are officially supported by both interfaces**.

Therefore, I propose to **reorganize and redesign Foreman's power operation UI and internal logic based on these shared functions**.

---

### List of Common Power Functions (IPMI / Redfish / virsh Comparison)

| UI Label               | IPMI Command              | Redfish Command                      | virsh Command         |
|------------------------|----------------------------|---------------------------------------|------------------------|
| **Power On**           | `power up (1h)`            | `ForceOn`, `On`                       | `start`                |
| **Force Power Off**    | `power down (0h)`          | `ForceOff`                            | `destroy`              |
| **Force Restart**      | `power cycle (2h)`         | `PowerCycle`, `FullPowerCycle`        | `reset`                |
| **Hard Reset**         | `hard reset (3h)`          | `ForceRestart`                        | `reset`                |
| **Safe Shutdown**      | `soft-shutdown via ACPI (5h)` | `GracefulShutdown`                | `shutdown`             |

---

### Mapping with Current Foreman Operations

The current Foreman power operations and their proposed mappings are shown below.

| Current Foreman Action | Proposed UI Label        | Notes                                                |
|------------------------|--------------------------|------------------------------------------------------|
| `Start`                | `Power On`               | Standard power-on operation                          |
| `Stop`                 | `Force Power Off`        | Immediate shutdown without notifying the OS          |
| `Poweroff`             | `Safe Shutdown`          | Graceful shutdown                |
| `Reboot`               | **Proposed for removal** | To be removed according to the direction of the original PR |
| `Reset`                | `Hard Reset`             | Sends a reset signal regardless of power state       |
| `Soft`                 | `Safe Shutdown`          | Equivalent to `Poweroff`; can be unified             |
| `Cycle`                | `Force Restart`          | Power-off followed by restart                        |

---

### Summary of Proposal

- **Redefine Foreman's power operations based on functionality officially supported by both IPMI and Redfish**
- **Remove or consolidate ambiguous or redundant actions such as `Reboot` and `Soft`**
- **Unify UI labels to be more intuitive and user-friendly**
- **Clearly map internal symbols to the respective IPMI, Redfish, and virsh commands**

---

### References

- [IPMI - Intelligent Platform Management Interface Specification Second Generation v2.0 (Section 28.3)](https://www.intel.la/content/dam/www/public/us/en/documents/specification-updates/ipmi-intelligent-platform-mgt-interface-spec-2nd-gen-v2-0-spec-update.pdf)
- [Redfish Data Model Specification v2025.1, Section 6.25.5.20](https://www.dmtf.org/sites/default/files/standards/documents/DSP0268_2025.1.pdf)

---

Looking forward to your feedback. Thank you!
